### PR TITLE
feat: csv to oscal cd reconcile2

### PR DESCRIPTION
**Pull request by degenaro**
## Types of changes

- [ ] Hot fix (emergency fix and release)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation (change which affects the documentation site)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [x] My code follows the code style of this project.
- [ ] Documentation for my change is up to date?
- [x] My PR meets testing requirements.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Key links:

- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

### CODE REVIEW

File: tests/trestle/tasks/csv_to_oscal_cd_test.py [hunk 1]
```diff
@@ -198,7 +206,7 @@ def _validate2(tmp_path: pathlib.Path):
     return
 
 
-def _validate3(tmp_path: pathlib.Path):
+def _validate03(tmp_path: pathlib.Path):
```
Comment (enikonovad) [inline tests/trestle/tasks/csv_to_oscal_cd_test.py:209]: What is `validate03`? What does it do? Please use a more explicit function name

File: tests/trestle/tasks/csv_to_oscal_cd_test.py [hunk 2]
```diff
@@ -268,10 +276,361 @@ def _validate3(tmp_path: pathlib.Path):
                 index_b].statement_id == smt[2]
 
 
+def _validate04(tmp_path: pathlib.Path):
```
Comment (enikonovad) [inline tests/trestle/tasks/csv_to_oscal_cd_test.py:279]: Ditto and to all other `validateXX` functions...

File: tests/trestle/tasks/csv_to_oscal_cd_test.py [hunk 3]
```diff
@@ -268,10 +276,361 @@ def _validate3(tmp_path: pathlib.Path):
...
+    # read catalog
+    fp = pathlib.Path(tmp_path) / 'component-definition.json'
+    cd = ComponentDefinition.oscal_read(fp)
+    # constants
+    ns = 'http://ibm.github.io/compliance-trestle/schemas/oscal/cd'
```
Comment (enikonovad) [inline tests/trestle/tasks/csv_to_oscal_cd_test.py:284]: https ?

File: tests/trestle/tasks/csv_to_oscal_cd_test.py [hunk 4]
```diff
@@ -268,10 +276,361 @@ def _validate3(tmp_path: pathlib.Path):
...
+    cd = ComponentDefinition.oscal_read(fp)
+    # spot check
+    assert len(cd.components) == 1
+    assert len(cd.components[0].control_implementations) == 1
+    assert len(cd.components[0].props) == 60
+    assert len(cd.components[0].control_implementations[0].implemented_requirements) == 21
+    assert len(cd.components[0].control_implementations[0].set_parameters) == 4
+
+
+def _validate18(tmp_path: pathlib.Path):
```
Comment (enikonovad) [inline tests/trestle/tasks/csv_to_oscal_cd_test.py:460]: validate 10,11,12,13,14,17,18 contain identical code except the number of different components. Please rewrite this.
 The same applies to all other `validateXX` functions. They are all highly similar and can be easily compressed...
 
 You can either use one function that takes those as parameters or use parameterized test (via `@pytest.mark.parametrize`)

File: tests/trestle/tasks/csv_to_oscal_cd_test.py [hunk 5]
```diff
@@ -363,11 +722,35 @@ def test_config_missing_csv_file(tmp_path: pathlib.Path):
...
+    """Test config missing user-namespace."""
+    config = configparser.ConfigParser()
+    config_path = pathlib.Path('tests/data/tasks/csv-to-oscal-cd/test-csv-to-oscal-cd.config')
+    config.read(config_path)
+    config.remove_option('task.csv-to-oscal-cd', 'user-namespace')
+    section = config['task.csv-to-oscal-cd']
+    tgt = csv_to_oscal_cd.CsvToOscalComponentDefinition(section)
+    retval = tgt.execute()
+    assert retval == TaskOutcome.FAILURE
+
```
Comment (enikonovad) [inline tests/trestle/tasks/csv_to_oscal_cd_test.py:747]: Again same here as above, these are duplicates of each other, same comment applies

File: trestle/tasks/csv_to_oscal_cd.py [hunk 6]
```diff
@@ -54,6 +55,13 @@ def derive_part_id(control_mapping: str) -> str:
     return rval
 
 
+def non_empty(list_: List) -> int:
```
Comment (enikonovad) [inline trestle/tasks/csv_to_oscal_cd.py:58]: Looking at the code it seems this function is used only to check whether the list is empty or not and if not then  append the list.
We have a function `as_list` that handles None and returns [] in such a case. Please use that instead

File: trestle/tasks/csv_to_oscal_cd.py [hunk 7]
```diff
@@ -72,14 +80,109 @@ def __init__(self, config_object: Optional[configparser.SectionProxy]) -> None:
...
+        self._timestamp = datetime.datetime.utcnow().replace(microsecond=0).replace(tzinfo=datetime.timezone.utc
+                                                                                    ).isoformat()
+        # config verbosity
+        self._quiet = self._config.get('quiet', False)
+        self._verbose = not self._quiet
+        # title
+        self._title = self._config.get('title')
+        if self._title is None:
+            logger.warning('config missing "title"')
+            return False
```
Comment (enikonovad) [inline trestle/tasks/csv_to_oscal_cd.py:144]: Compress lines 141 to 164 to a single call
```
load_config():
   for c in ['title', 'version', 'namespace',......]:
        if c is None:
             logger.warning(f'config missing {c}')
             return False
```

File: trestle/tasks/csv_to_oscal_cd.py [hunk 8]
```diff
@@ -112,262 +212,1207 @@ def _execute(self) -> TaskOutcome:
...
+        """Control mappings modify."""
+        # Not currently possible
+
+    def user_columns_add(self) -> None:
+        """Add user columns."""
+        # Future
+
+    def user_columns_del(self) -> None:
+        """Delete user columns."""
+        # Future
```
Comment (enikonovad) [inline trestle/tasks/csv_to_oscal_cd.py:759]: Delete those for now, one time comes those can be rewritten

**Review by enikonovad — CHANGES_REQUESTED** [review]
Some major changes needs to be done here:
#### Test data
Test data contains a lot of duplicates, brief look at bp.sampleX.csv showed that it differs but a couple of lines/words.
It is really hard to manage so many test files.
Please instead have a few test files that contain data and then use code to modify that data to suit your needs. 

#### Code duplication
There is a lot of code duplication. Please see comments. A lot of code can be compressed to a single function. Managing code that differs only by a couple of words is hard as it requires you to modify it in N places instead of one if something changes.

**Comment by degenaro** [comment]
@enikonovad Thx for the review! All your comments/requests are good and reasonable. I'll work on the needed changes presently...

File: trestle/tasks/csv_to_oscal_cd.py (hunk 8 above)
Comment (degenaro) [reply trestle/tasks/csv_to_oscal_cd.py:759]: I understand clutter avoidance, though we do lose information by removing.

File: trestle/tasks/csv_to_oscal_cd.py (hunk 6 above)
Comment (degenaro) [reply trestle/tasks/csv_to_oscal_cd.py]: Fixed.

File: trestle/tasks/csv_to_oscal_cd.py (hunk 8 above)
Comment (degenaro) [reply trestle/tasks/csv_to_oscal_cd.py]: Fixed.

**Comment by degenaro** [comment]
@enikonovad I'm withdrawing this PR in favor of https://github.com/IBM/compliance-trestle/pull/1272. I thought it was easier to just create a new PR than to fix-up this one. I claim to have addressed the changes you requested in the new PR, most especially in the reduction of test data files (in fact, reducing even the original test data significantly). The trestle task code itself is largely unchanged from the previous PR, modulo the few changes you requested. I'll leave this PR alive for now as reference, but change to draft status.

**Comment by degenaro** [comment]
Abandon in favor or PR #1272 .

**Pull request closed without merging**